crates/astria-sequencer/src/mempool.rs

```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
     }
 }
 
+const TX_TTL: i64 = 600; // 10 minutes 
+const INVALID_CACHE_SIZE: usize = 4096;
+const TX_INVALID_TTL: i64 = 60; // 60 seconds
+
+/// `InvalidCache` provides for signaling to `CometBFT` what transactions
+///  can be removed from the `CometBFT` mempool.
+///
+/// The cache places a TTL on the entries, allowing users
+/// to request their transaction to be processed again after the
+/// ttl has expired. This is useful for users who want to re-run a transaction
+/// that failed to execute or to re-add a transaction that has expired.
+#[derive(Clone)]
+pub(crate) struct InvalidCache {
+    cache: HashSet<[u8; 32]>,
+    time_added: HashMap<[u8; 32], i64>,
+    remove_queue: VecDeque<[u8; 32]>,
+    max_size: usize,
+    time_to_live: i64,
+    size: usize,
+}
+
+impl InvalidCache {
+    fn new(max_size: usize, time_to_live: i64) -> Self {
+        Self {
+            cache: HashSet::new(),
+            time_added: HashMap::new(),
+            remove_queue: VecDeque::with_capacity(max_size),
+            max_size,
+            time_to_live,
+            size: 0,
+        }
+    }
+
+    fn cached(&self, tx_hash: [u8; 32]) -> bool {
+        // the tx is known and entry hasn't expired
+        self.cache.contains(&tx_hash)
+            && (Time::now().unix_timestamp()
+                <= self.time_added[&tx_hash]
+                    .checked_add(self.time_to_live)
+                    .expect("overflowed ttl add"))
+    }
+
+    fn add(&mut self, tx_hash: [u8; 32]) {
+        if self.cache.contains(&tx_hash) {
+            // update time to live if already exists
+            // note: this doesn't change the tx's position in the remove vector
+            self.time_added
+                .insert(tx_hash, Time::now().unix_timestamp());
+            return;
+        };
+        if self.size == self.max_size {
+            // make space for the new transaction by removing the oldest transaction
+            let removed_tx = self
+                .remove_queue
+                .pop_front()
+                .expect("cache should contain elements");
```

Comment:
I don't think we need to panic here - maybe an error log if we get `None`?

Alternatively, if we changed `max_size` to be a `std::num::NonZeroUsize` and changed `if self.size == self.max_size` to `if self.remove_queue.len() == usize::from(self.max_size)` it would be reasonable to leave the `expect` since it could never be hit.